src/parser/core/modules/EntityStatuses.ts

```diff
@@ -0,0 +1,142 @@
+import STATUSES from 'data/STATUSES'
+import {BuffEvent} from 'fflogs'
+import Module, {dependency} from 'parser/core/Module'
+import {getDataBy} from '../../../data'
+import {FFLogsEventNormaliser} from './FFLogsEventNormaliser'
+import Invulnerability from './Invulnerability'
+
+const APPLY = 'apply'
+const REMOVE = 'remove'
+
+interface BuffTrackingEvent extends BuffEvent {
+	start: number,
+	lastRefreshed: number,
+	end: number | null,
+	stacks: number,
+	stackHistory: StackHistoryEvent[],
+	isDebuff: boolean,
+}
+
+interface StackHistoryEvent {
+	stacks: number,
+	timestamp: number,
+	invuln: boolean,
+}
+
+interface StatusRangeEndpoint {
+	timestamp: number,
+	type: 'apply' | 'remove',
+	stackHistory: StackHistoryEvent[],
+}
+
+export class EntityStatuses extends Module {
+	static handle = 'entityStatuses'
+	static debug = false
+
+	@dependency private invuln!: Invulnerability
+	@dependency private fflogsEvents!: FFLogsEventNormaliser
+
+	// -----
+	// API
+	// -----
+	// TODO: This implementation which I've shamelessly stolen seems to only track overall
+	//       uptime, ignoring potential gains from multidotting, etc. Should that be in here,
+	//       or is that a somewhere-else sort of thing...
+	getStatusUptime(statusId: number, onTargets: TODO, sourceId = this.parser.player.id) {
+		// Search for status maps on valid targets
+		const statusEvents: BuffTrackingEvent[] = Object.values(onTargets)
+			.reduce((statusEvents: BuffTrackingEvent[], target: TODO) => {
+				const appliedStatuses = target.buffs?.filter((statusEvent: BuffTrackingEvent) => {
```

Comment:
that `?.buffs` has a potential to return `undefined`, which JS'll happily concat onto the array as a new item. You'll probably need to return the accumulator early or something similar.

Not sure why TS isn't complaining about it; best guess is that the `any` from the `TODO` is squashing the `| undefined` that `?.` should generate.